src/main/scala/akka/persistence/cassandra/query/journal/QueryActorPublisher.scala

```diff
@@ -0,0 +1,129 @@
+package akka.persistence.cassandra.query.journal
+
+import scala.concurrent.duration.FiniteDuration
+import scala.concurrent.{ExecutionContext, Future}
+
+import akka.actor.{Actor, ActorLogging}
+import akka.pattern.pipe
+import akka.stream.actor.ActorPublisher
+import akka.stream.actor.ActorPublisherMessage.{Cancel, Request, SubscriptionTimeoutExceeded}
+
+//TODO: Optimizations - manage the buffer size efficienly, e.g. based on nr requests, remaining elements etc.
+//TODO: Database timeout, retry and failure handling.
+//TODO: Write tests for buffer size, delivery buffer etc.
+/**
+ * Abstract Query publisher. Can be integrated with Akka Streams as a Source.
+ * Intended to be extended by concrete Query publisher classes. This class manages the stream lifecycle,
+ * live stream updates, refreshInterval, max buffer size and causal consistency given an
+ * offset queryable data source. Causality is achieved by only a single request for data in flight
+ * at any point in time. Implementer must provide concrete initial state, update operation
+ * and end condition.
+ *
+ * @param refreshInterval Refresh interval.
+ * @param maxBufferSize Maximal buffer size.
+ * @tparam MessageType Type of message.
+ * @tparam State Type of state.
+ */
+private[journal] abstract class QueryActorPublisher[MessageType, State](
```

Comment:
The purpose of this class is to abstract stream publishing logic. It handles all required aspects - refresh interval, from and to offsets, buffer size, live stream updates etc. The implementing classes only worry about database queries and buffer and state update logic (e.g. causality etc.).

The reason behind this was to separate concerns. Stream state essentially is a (very, very loosely :)) a monoid like structure. We have a zero (starting state) and a single operation that updates the state when data are received. In addition to that we define the query as (State, Max) => Future[Vector[T]] and completion condition as State => Boolean (e.g. toOffset reached). This should be sufficiently powerful to express all query publishers.

The implementation is imo quite clear, using completely immutable data. State is only managed by updates to the behavior stack and most if not all functions are pure. It should therefore be simple to reason about the state transitions and behaviour, including concurrency.